cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/TetherServerHandler.java

```diff
@@ -0,0 +1,273 @@
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.TopicId;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.HandlerContext;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+
+/**
+ * {@link io.cdap.http.HttpHandler} to manage tethering server v3 REST APIs
+ */
+@Path(Constants.Gateway.API_VERSION_3)
+public class TetherServerHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(TetherServerHandler.class);
+  private static final Gson GSON = new GsonBuilder().create();
+  private static final String SUBSCRIBER = "tether.server";
+  static final String TETHERING_TOPIC_PREFIX = "tethering_";
+  private final CConfiguration cConf;
+  private final TetherStore store;
+  private final MessagingService messagingService;
+  private final MultiThreadMessagingContext messagingContext;
+  private final TransactionRunner transactionRunner;
+  // Last processed message id for each topic. There's a separate topic for each peer.
+  private Map<String, String> lastMessageIds;
+
+  @Inject
+  TetherServerHandler(CConfiguration cConf, TetherStore store, MessagingService messagingService,
+                      TransactionRunner transactionRunner) {
+    this.cConf = cConf;
+    this.store = store;
+    this.messagingService = messagingService;
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.transactionRunner = transactionRunner;
+  }
+
+  @Override
+  public void init(HandlerContext context) {
+    super.init(context);
+    initializeMessageIds();
+  }
+
+  /**
+   * Sends control commands to the client.
+   */
+  @GET
+  @Path("/tethering/controlchannels/{peer}")
+  public void connectControlChannel(HttpRequest request, HttpResponder responder, @PathParam("peer") String peer)
+    throws IOException, NotImplementedException, PeerNotFoundException {
+    checkTetherServerEnabled();
+
+    store.updatePeerTimestamp(peer);
+    TetherStatus tetherStatus = store.getPeer(peer).getTetherStatus();
+    if (tetherStatus == TetherStatus.PENDING) {
+      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
```

Comment:
Prefer to throw exception instead of send status and return.

Reply:
fixed